networkx/algorithms/tournament.py

```diff
@@ -318,6 +335,62 @@ def is_reachable(G, s, t):
            <http://eccc.hpi-web.de/report/2001/092/>
     """
 
+    try:  # Use numpy if available, otherwise fall back to pure Python implementation
+        return _is_reachable_numpy(G, s, t)
+    except ImportError:
+        return _is_reachable_python(G, s, t)
+
+
+def _is_reachable_numpy(G, s, t):
+    import numpy as np
+
+    def two_neighborhood(adj_matrix, v, node_indices):
+        """Returns the set of nodes at distance at most two from `v`.
+
+        `adj_matrix` must be a numpy 2D boolean array representing the
+        adjacency matrix of the graph, `v` is the index of a node in the graph,
+        and `node_indices` is an iterable of all node indices in the graph.
+
+        The returned set includes the nodes at distance zero (that is,
+        the node `v` itself), the nodes at distance one (that is, the
+        out-neighbors of `v`), and the nodes at distance two.
+
+        """
+        return {
+            x
+            for x in node_indices
+            if x == v
+            or adj_matrix[v, x]
+            or any(adj_matrix[v, z] and adj_matrix[z, x] for z in node_indices)
+        }
+
+    def is_closed(adj_matrix, nodes, node_indices):
+        """Decides whether the given set of nodes is closed.
+
+        A set *S* of nodes is *closed* if for each node *u* in the graph
+        not in *S* and for each node *v* in *S*, there is an edge from
+        *u* to *v*.
+
+        """
+        return all(adj_matrix[u, v] for u in set(node_indices) - nodes for v in nodes)
+
+    nodelist = list(G)
+    adj_matrix = nx.to_numpy_array(G, dtype=bool, nodelist=nodelist)
+    node_map = {node: i for i, node in enumerate(nodelist)}
+    node_indices = range(adj_matrix.shape[0])
+    s = node_map[s]
+    t = node_map[t]
+
+    neighborhoods = [
+        two_neighborhood(adj_matrix, v, node_indices) for v in node_indices
+    ]
+    return all(
+        not (is_closed(adj_matrix, S, node_indices) and s in S and t not in S)
+        for S in neighborhoods
+    )
```

Comment:
Would it make sense to only compute neighborhoods as they're evaluated? Right now we're computing all of them upfront when the first one might already decide the returned value.